### ISSUE TITLE
Full parity w/ Java Mode's PVector class

### ISSUE THREAD

**Issue by GoToLoop** [processing-js/processing-js#200]
Just finished it & haven't tested it & dunno how to either:
https://Gist.GitHub.com/GoToLoop/acbf106aa784820aff23  
It also expands upon https://GitHub.com/processing-js/processing-js/pull/197
And fixes: https://github.com/processing-js/processing-js/issues/199  
Some spotlights below:

1. Fully chainable methods for all instances that previously didn't return anything.
2. Optional _target_ parameter to store the result. So we can avoid redundant _**PVector**_ instantiations.
3. Static calls like `PVector.dist(vec1, vec2);` can also use any _**PVector**_ reference just like Java: `vec1.dist(vec1, vec2);`
4. Included both **equals()** & **hashCode()** just like Java Mode.
5. Added **add()**, **mult()** & **div()** as static method calls as well.
6. Removed non-compliant behavior like `this.x *= v.x;` from **mult()** & **div()** which doesn't exist in Java Mode API.

**Comment by Pomax** [comment]
Thanks!

Testing is relatively straight forward: the `./test/unit` directory contains a number of pde files that all test individual aspects of the code. The main PVector test is `PVector.pde`, so have a look at that; it should be relatively straight forward to add checks based on the code you've written that fail without your code, but pass with it. You can check that your JS passes linting with `$> grunt`, and you can run your code against the unit tests using `$> node test`. 

If you turn the gist into a pull request instead, it can be code reviewed and landed in: if you haven't already, fork this repository to your account, then create a new branch for this fix, replace the file you edited with the code in your gist, and then tell github to create a pull request onto the processing-js repository.

If you are uncomfortable with testing, you can even do all of that online (although then it's harder to know whether you actually got everything right).

A few comments on the gist, since there is PR to leave line comments on yet:

- There is nothing in [the Processing docs](https://processing.org/reference/PVector_get_.html) to indicate `get` is deprecated, and there is no `copy` listed in the official API, so let's take that back out.
- let's not define `lerp` again, it's already defined in too many places. We can probably pull it in from any of a number of sources that already carry it.
- JS rarely needs `instanceof`, we can just directly test for the expected function and if it's there, use it, if not fall back to a safety instead.
- As ES5 flavoured JavaScript, `const` doesn't actually do anything. let's turn all of them into `var` instead.
- What do you mean with "non-compliant behavior"? Remember that Processing.js is about parity, **not** "having the same code" so the only thing that has to be compliant is parameter acceptance and function results for correct code. Bad calls, as well as function internals/implementation details don't need to be the same in the slightest (see the [contributors.md](https://github.com/processing-js/processing-js/blob/master/CONTRIBUTING.md#the-fundamental-rule-for-processingjs) section on parity for more information on that)
- `argsErr` is a good idea, but let's also make it say which function it failed in. Having the throw gives a trace, but immediately seeing which function that occurred in cuts down the debug time.
```
        argsErr = function(fname, len, min) {
          throw 'too few arguments passed to ' + fname + ' (' + len + ' < ' + min + ')';
        };
```

**Comment by GoToLoop** [comment]
> If you turn the gist into a pull request instead, it can be code reviewed and landed in...

For now this is just a preview for this Processing's 100% refactoring & parity for _**PVector**_ class.
My intention is that it can be seen in its entirely w/o any diff. comparisons getting in the way.
After all, I've even reshifted where methods are placed within the code!
Once it's more formely approved, I'm gonna issue a pull request for it. No worries!

> If you are uncomfortable with testing, ...

It's more than uncomfortable: I dunno how to do it at all!
Most of the time I just use Processing's IDE (PDE).
However for this particular "gist", I've coded it using Atom IDE.
I just hope some1 else w/ more experience can merge it within PJS and check it out live.

> There is nothing in the Processing docs to indicate **get()** is deprecated,
and there is no **copy()** listed in the official API, so let's take that back out.

I can't do that for it would violate 1 of the main reasons I've done it all: 100% parity w/ Processing!
As 1 of Processing forum's main answerers, dealing w/ users every single day, I also keep close tabs w/ any Processing API's changes.
And I remember well that **copy()** has been landed in the middle of Processing 2's "season".
Aiming to replace **get()** in returning a `new` _**PVector**_ instance w/ the same values for _x_, _y_ & _z_.

About Java Mode's web reference, it's always outdated, fulla errors & omissions!
Be aware that **get()** overloaded version w/ _target_ parameter _**isn't**_ deprecated. Only the empty **get()** is.

And it is the "PVector.java" itself which says so for both **get()** & **heading2D()**:
https://github.com/processing/processing/blob/master/core/src/processing/core/PVector.java#L375
https://github.com/processing/processing/blob/master/core/src/processing/core/PVector.java#L905

Actually I've found the "invention" of **copy()** very silly. I'm pretty comfortable w/ **get()** btW!
And of course I haven't removed either **get()** nor **heading2D()**. Merely commented them as `@Deprecated`:

```js
    get: function(target) {
      if (!arguments.length) { return this.copy(); } // @Deprecated
```

```js
  PVector.prototype.heading2D = PVector.prototype.heading; // @Deprecated
```

For 1 more "hard" proof that Processing's devs indeed added _**PVector**_'s **copy()** & deprecated its **get()**:
None other than @shiffman has entirely removed **get()**'s alias option for p5.js project:
https://github.com/processing/p5.js/pull/506

> Let's not define **lerp()** again. It's already defined in too many places.

I don't get why would you "accuse" me of that! Since it is already presently there as **lerp_val()**:
https://github.com/processing-js/processing-js/blob/master/src/Objects/PVector.js#L184

```js
  lerp: function(v_or_x, amt_or_y, z, amt) {
      var lerp_val = function(start, stop, amt) {
        return start + (stop - start) * amt;
      };
```

I've merely relocated it to the top of the file and renamed it as **lerp()**.
Now it lives on as an internal private closure. Available solely to the _**PVector**_ class.
I've done that in order to skip reinstantiation of it every single time **lerp()** method is invoked!

I really wished I had access to _**Processing__*'s Math API inside _**PVector**_. So I coulda imported both **lerp()** & **sq()**.
But alas, only _**PConstants**_ are passed to _**PVector**_ module. That's where I've got TAU:

```js
  const TAU  = options.PConstants.TAU,
```

We gotta work w/ what's available for us after all. Although I coulda just defined TAU as `2*Math.PI` instead.

> JS rarely needs `instanceof`, we can just directly test for the expected function...

As mainly a Java programmer, I'm very used to check by nominal datatype only. So things like `instanceof` & `typeof` are natural to me.
But you're correct that in JS we can check structurally instead. Whether some properties are present or not in some object.

Therefore I can replace this nominal check for class _**Processing**_:
```js
    const rnd = parent? parent : target instanceof Processing? target : Math;
```

W/ this structural check for property _random_:
```js
    const rnd = parent? parent : target && target.random? target : Math;
```

Since **random()** is the only method both **PVector.random2D()** & **PVector.random3D()** are interested in!
Thus I've changed those 2 methods to use this more flexible approach you've tipped about instead. :D

However for the rest, where `instanceof PVector` is being checked, it's not feasible to shift them to structure check.
They need all _x_, _y_ & _z_ properties in addition to any methods they might need to exist! So I've left them unchanged.

> Let's turn all of them into `var` instead.

For the time being I'd rather prefer to leave those `const` declarations in the source code.
When it's about ready to merge, it's just as easy as replace all "const " w/ "var " and fix indentations 2 spaces to left.

FYI, even though `const` keyword is ECMA 6, it's been recognizable for all main browser families for a long time now.
The only biggest exception is IE as always. Only its latest IE11 recognizes its existence!  =(

I dunno how much more you wanna keep compatibility w/ such old IE versions like IE10, IE9 & IE8, etc.
But once you have IE11 as minimum version, `const` can be used w/o any worries!
Hopefully `let` as well till then! (^_^)

> What do you mean with "non-compliant behavior"? Remember that Processing.js is about parity...

I mean that 1 of the overloaded operations doesn't exist in _**PVector**_ class. Perhaps it _**never**_ existed! Not sure...
Let's compare web references for **mult()** for both Java & JS modes:
http://Processing.org/reference/PVector_mult_.html
http://ProcessingJS.org/reference/PVector_mult_/

While Java Mode cites 3 overloaded syntaxes: `.mult(n)`, `.mult(v, n)`, `.mult(v, n, target)`.
JS Mode cites an extra 1 though: `.mult(v1,v2)`.

I've just removed the extra overload behavior b/c I thought you would demand that!
Y'know, I'm all in favor w/ new features and high flexibility for API.
If you allow it, I'd gladly put them back both in **mult()** & **div()**!

> Having the `throw` gives a trace, but immediately seeing which function that occurred in cuts down the debug time.

You're right! I thought JS would automatically create a stack trace a la Java and point to the faulty method. Silly me!
Implemented your solution there now!  >;-)

**Comment by Pomax** [comment]
A PR makes it easier to leave comments directly where they are relevant, so one way to make sure there's no "ugly diffing" is to give the file a new name, with a deletion of the old one, and then settup up PVector with a `require('newname')`. All the way at the end, when the code is deemed "done", it's then trivial to set up one more commit to rename it back to the original name and land it.

It also makes it possible for others to do the testing, if you don't know how to run them (I'll have to write some more docs on how to do a full PR, contributors.md currently doesn't explain this well enough, and the readme.md certainly doesn't!), since a PR can be pulled in, whereas `git` doesn't have any built in commands for pretending that a gist is a specific file that can be used to replace an existing one, with a commit hash to allow rollback.

Also, an important note here: review code comments are always about the code, not the coder. Specifically in this case, if the class gets rewritten then it is worth taking out code that doesn't belong in PVector, even though it's currently found in the PVector file, especially when it means removing code that really shouldn't have been there from the get-go: `lerp` is a great example of that. I'm not accusing you of putting it in, I am pointing out that while the current implementation has its own copy of that, it *really* shouldn't, since we already have it implemented in several places. So instead, it'll be far better to simply reuse one of those implementations instead of maintaining the level of non-[DRY](https://en.wikipedia.org/wiki/Don't_repeat_yourself)-ness that the current code suffers from.

I also agree it would be useful to have access to the Math object, although many of the Processing.Math implementations are aliassed to JavaScript's native Math object. Especially for Math operations it tends to be worth not relying on Processing.Math but to just use Math directly for the thing it is needed for, to cut out of the branching overhead. It's been the cause of perf issues in the past, so keeping that part of the code lean tends to pay off.

Pjs currently only supports the Processing 2.0 API, so this uplift that you're doing is super valuable, getting parts to 3.0 will help a lot of people, and the main criterium for landing code into Pjs is "does it improve parity with the current API". Extending the functions to accept 3.0 call signatures is a perfect example of that. I'm still not sure what you meant with non-compliant behaviour, since the example you gave is `this.x *= v.x;`, which is a line of code inside a function, which makes it an implementation detail and irrelevant in terms of changes: if you need to change a function's internals so that it supports one or more current signatures, then that is a matter of course.

As a last note on stack traces: JS does generate stack traces, but throws are passed around a fair bit inside Pjs so you may not necessarily see a trace "end" in the function that threw. That's why it's generally a good idea to also have the throw message itself contain all the information needed to home in on where things went wrong.

Thanks again for all the work!

**Comment by GoToLoop** [comment]
Internal **lerp()** helper is called in this code section inside _**PVector**_'s **lerp()** method:
```js
      this.x = lerp(this.x, x, amt);
      this.y = lerp(this.y, y, amt);
      this.z = lerp(this.z, z, amt);
```
I don't get why are you so upset about having **lerp()** helper as a private closure for _**PVector**_'s **lerp()**.
It's in the same league as **argsErr()**, **sq()** & TAU! Either that or we need to import _**Processing**_'s own **lerp()**!
And no, JS' _**Math**_ doesn't have **lerp()**. It is Processing's own API implementation and not a wrapper call.

Even Java Mode relies on **PApplet.lerp()** inside its own _**PVector**_'s **lerp()**:
https://github.com/processing/processing/blob/master/core/src/processing/core/PVector.java#L947

Of course I coulda transposed **lerp()**'s formula in situ as:
`this.x = lerp(this.x, x, amt);` -> `amt * (x - this.x) + this.x`

But IMO, it's not as succinct nor more DRY!  :P

**Comment by Pomax** [comment]
Again you for some reason that I do not understand think that I am somehow emotionally affected here. Please understand that "commenting on code" has absolutely nothing to do with "emotional state": one of the most important parts of writing any code is to make sure there is as little repeated code as possible, because any code duplication risks potential divergence and thus potential bugs. When writing any large library, one of the main guidelines is to try to not repeat yourself whenever you can, if there is code that is used in two different places. This is one of those instances, so let's try to avoid repeating code, to keep divergence down as much as possible. Anything *else* we can further remove from this code because we can reuse it from somewhere else is *also* an excellent improvement, unless it adversely affects performance, in which case it's not and we should keep them local to the file with a note that we're doing so for performance reasons.

**Comment by GoToLoop** [comment]
I too adhere to DRY approach! But you still hadn't pointed out which course to take for **lerp()**:

1. Keep the formula as a _**PVector**_'s closure helper function just like I did for **sq()**.
2. Leave it as a local function inside method **lerp()** as it is currently implemented.
3. Import it from Processing's _**Math**_ library like Java Mode did in the form of **PApplet.lerp()**.
4. Repeat the formula in situ 3 times for _this.x_, _this.y_ & _this.z_.

**Comment by Pomax** [comment]
Right, so one thing I think I forgot to point out here (because I assumed it was known, but that's a bad assumption) is that Processing.js does not follow the "vanilla" client side javascript model, but is written using the Node.js version of CommonJS. This means we have import functionality available, which in turn means that getting things "from somewhere else" is actually really easy.

We can get lerp and TAU and everything else that is already defined somewhere is into this code with some really easy syntax:
```
var TAU = require("./relative/location/of/PConstants").TAU.
    lerp = require("./relative/location/of/Math").lerp,
    // etc.
```
Which is rather nice: it gives local aliasses, but keeps everything only declared once.

There's quite a few places where this was not taken advantage of during the initial port from vanilla JS to CommonJS format, so I'll see if I can find all the spots and open issues for each of those.

**Comment by GoToLoop** [comment]
Actually I was hoping to discuss Java Mode & PJS implementation differences like:
```js
    heading: function() {
      //return -Math.atan2(-this.y, this.x);
      return Math.atan2(this.y, this.x);
    }
```
Which 1 is more correct? PJS' `-Math.atan2(-this.y, this.x)` or P5's `Math.atan2(this.y, this.x)`?

In **angleBetween()**:
```js
  PVector.angleBetween = function(v1, v2) {
    if (!v1.x && !v1.y && !v1.z || !v2.x && !v2.y && !v2.z) { return 0; }
    const amt = PVector.dot(v1, v2) / Math.sqrt(v1.magSq() * v2.magSq());
    return amt <= -1? Math.PI : amt >= 1? 0 : Math.acos(amt);
  };
```
Is there any chance that the result _amt_ can be <= -1 or >= 1?
A fixed made by @viamodulo thinks `Math.sqrt(v1.magSq() * v2.magSq())` is enough to avoid NaN: https://github.com/processing-js/processing-js/pull/169
Or should I follow the extra checks like Java Mode worries about?

Is my **hashCode()** adaptation reasonable for JS, even though the hash isn't an integer?:
```js
    hashCode: function() {
      var hash = 1;
      hash = 31*hash + x;
      hash = 31*hash + y;
      return 31*hash + z;
    }
```

**Comment by Pomax** [comment]
neither of them is better, the official function signature of `atan2` is that it takes `(dy, dx)`, but the only proper answer is "do the outputs match". If this is a user-facing function (so it's in the Processing reference, not just a P5 internal helper function that happens to be `public` because that's what it has to be for P5 to be able to run its code), then Pjs should have the same input->output relation as P5.

If, however, it's a purely internal function, not exposed as part of the public API, then it doesn't matter, and the adage "if it works, it works" applies: don't fix it until it clearly breaks something. File it as a new issue to examine whether or not to clean it up, and then ignore it until you (or someone else) works on that issue.

As for the hash, hashCodes have the lovely property that "hashcodes are in no way intended to be the same across different implementations". If your hashcode missors the java approach, then that's fine. Since all the operations in that code are integer operations (no division) the result will still be integer, although if you want to force 32 bit integers, you can do this:

```
    hashCode: function() {
      var hash = 1;
      hash = (31*hash + x) | 0;
      hash = (31*hash + y) | 0;
      return (31*hash + z) | 0;
    }
```
The `... | 0` is a bitwise operation, which in JS cast all input to fit a 32 bit integer. You can also leave it as is, JS has perfect integer representation for +/- 2^(53)-1, which we won't run into unless the coordinates are stupendously huge (ranging in the hundreds of billions). So either is fine, the only thing that matters is that *during a single run of a program*, the hashcodes are stable.

**Comment by GoToLoop** [comment]
> If this is a user-facing function...

I've got the impression you hadn't checked the code enough. B/c even after all this discussion you hadn't realized only **lerp()**, **sq()** & **argsErr()** are private closure functions. Alas, I just wanted more direct answers.
Since quality progression is no longer possible, it's better to issue a pull request the way it is now. :-(

**Comment by Pomax** [comment]
Yes, that "if" lived in a block that explains how developing for Processing.js works.

If it's (where "it" is **any** function) a public API function, then Processing.js needs to behave the same as Processing.

If it's not (where "it" is still **any** function) a public API function, then we don't care.

**heading()** is a public function, it's documented on https://processing.org/reference/PVector_heading_.html, and so given the base guidelines stated, Pjs needs to do the same thing as P5. So that would have been a reasonable response:

"It's a public function, and I tested it for some inputs but Pjs generates the wrong output, so we should change it". Cool, had you said that, then we're done: there is proof Pjs is doing the wrong thing, and we need to fix that.

As for the other three functions, **lerp** should be a require() call, and if that's too difficulat because it's bound as `p.lerp` in `src/P5Functions/Math.js` then we need to give it its own file like `lerp.js` with content
```
module.exports = function(r, v1, v2) {
  return ((v2 - v1) * r) + v1;
};
```
and then `require()` that file both in PVector.js, and in Math.js, so that we only have it declared in one, authoritative, place.

**sq** is the *perfect* example of what I've been talking about: you need this function to make your code work, it makes the code easier to read, **awesome, keep that function in**. And if any other code needs this function too, then just like anything else let's refactor so it's declared only once, and everything that needs it `require()`s it in.

**argsErr** I have no idea why this is still confusing, I literally said this was a good idea, as long as we also add which function things go wrong in. This will be a great candidate for refactoring into its own module later, too, but right now only PVector will be using it so there is no reason to refactor yet.
